### ISSUE TITLE
build will sometimes fail due to scikit-build caching directory `_skbuild` which does not support `pip --build-isolation`

### ISSUE THREAD

**Issue by joakim-hove** [equinor/ert#2408]
Running `pip install -e .` fails

**To Reproduce**
1. Clean checkout
2. Pristine virtual env
3. `pip install -r dev-requitements.txt`
4.  `pip install -e .`

**Expected behavior**
Happy build

**Traceback**
```
  ERROR: Command errored out with exit status 1:
     command: /home/hove/work/ERT/venv/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/hove/work/ERT/ert/setup.py'"'"'; __file__='"'"'/home/hove/work/ERT/ert/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
         cwd: /home/hove/work/ERT/ert/
    Complete output (72 lines):
    /tmp/pip-build-env-ic_sq90r/overlay/lib/python3.9/site-packages/setuptools/command/easy_install.py:156: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
      warnings.warn(
    Not searching for unused variables given on the command line.
    -- The C compiler identification is GNU 10.2.1
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Check for working C compiler: /usr/bin/cc - skipped
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- The CXX compiler identification is GNU 10.2.1
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Check for working CXX compiler: /usr/bin/c++ - skipped
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    -- Configuring done
    -- Generating done
    -- Build files have been written to: /home/hove/work/ERT/ert/_cmake_test_compile/build
    CMake Error at CMakeLists.txt:2 (project):
      Running
    
       '/tmp/pip-build-env-oxm_mcmn/overlay/bin/ninja' '--version'
    
      failed with:
    
       No such file or directory
    
    
    -- Configuring incomplete, errors occurred!
    See also "/home/hove/work/ERT/ert/_skbuild/linux-x86_64-3.9/cmake-build/CMakeFiles/CMakeOutput.log".
    See also "/home/hove/work/ERT/ert/_skbuild/linux-x86_64-3.9/cmake-build/CMakeFiles/CMakeError.log".
      File "/tmp/pip-build-env-ic_sq90r/overlay/lib/python3.9/site-packages/skbuild/setuptools_wrap.py", line 586, in setup
        env = cmkr.configure(cmake_args,
      File "/tmp/pip-build-env-ic_sq90r/overlay/lib/python3.9/site-packages/skbuild/cmaker.py", line 237, in configure
        raise SKBuildError(
    An error occurred while configuring with CMake.
      Command:
        cmake /home/hove/work/ERT/ert/libres -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/home/hove/work/ERT/ert/_skbuild/linux-x86_64-3.9/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/home/hove/work/ERT/venv/bin/python3 -DPYTHON_VERSION_STRING:STRING=3.9.2 -DPYTHON_INCLUDE_DIR:PATH=/usr/include/python3.9 -DPYTHON_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpython3.9.so -DSKBUILD:INTERNAL=TRUE -DCMAKE_MODULE_PATH:PATH=/tmp/pip-build-env-ic_sq90r/overlay/lib/python3.9/site-packages/skbuild/resources/cmake -DRES_VERSION=2.30.0b1.dev10+gc2f063430 -DECL_INCLUDE_DIRS=/tmp/pip-build-env-ic_sq90r/overlay/lib/python3.9/site-packages/ecl/.include -DCMAKE_OSX_DEPLOYMENT_TARGET=10.15 -DCMAKE_INSTALL_LIBDIR=res/.libs -DCMAKE_BUILD_TYPE:STRING=Release
      Source directory:
        /home/hove/work/ERT/ert/libres
      Working directory:
        /home/hove/work/ERT/ert/_skbuild/linux-x86_64-3.9/cmake-build
    Please see CMake's output for more information.
    
    
    --------------------------------------------------------------------------------
    -- Trying "Ninja" generator
    --------------------------------
    ---------------------------
    ----------------------
    -----------------
    ------------
    -------
    --
    --
    -------
    ------------
    -----------------
    ----------------------
    ---------------------------
    --------------------------------
    -- Trying "Ninja" generator - success
    --------------------------------------------------------------------------------
    
    Configuring Project
      Working directory:
        /home/hove/work/ERT/ert/_skbuild/linux-x86_64-3.9/cmake-build
      Command:
        cmake /home/hove/work/ERT/ert/libres -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/home/hove/work/ERT/ert/_skbuild/linux-x86_64-3.9/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/home/hove/work/ERT/venv/bin/python3 -DPYTHON_VERSION_STRING:STRING=3.9.2 -DPYTHON_INCLUDE_DIR:PATH=/usr/include/python3.9 -DPYTHON_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpython3.9.so -DSKBUILD:INTERNAL=TRUE -DCMAKE_MODULE_PATH:PATH=/tmp/pip-build-env-ic_sq90r/overlay/lib/python3.9/site-packages/skbuild/resources/cmake -DRES_VERSION=2.30.0b1.dev10+gc2f063430 -DECL_INCLUDE_DIRS=/tmp/pip-build-env-ic_sq90r/overlay/lib/python3.9/site-packages/ecl/.include -DCMAKE_OSX_DEPLOYMENT_TARGET=10.15 -DCMAKE_INSTALL_LIBDIR=res/.libs -DCMAKE_BUILD_TYPE:STRING=Release
    
    Traceback (most recent call last):
    
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/hove/work/ERT/venv/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/hove/work/ERT/ert/setup.py'"'"'; __file__='"'"'/home/hove/work/ERT/ert/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
```

**Enviromment**
 - OS: Debian stable
 - Python version: 3.9.2

**Comment by dotfloat** [comment]
You need to remove the `_skbuild` directory. Really annoying but there doesn't seem like there's anything we can do other than switch to some other build system.

**Comment by jondequinor** [comment]
Is this due to build isolation? Maybe `--no-build-isolation` fixes this?

scikit creates the _skbuild dir then pip introduces build isolation by creating a temp folder (where the ninja overlay is placed) and this works fine. However, the _skbuild dir persists, but the isolated part of the build does not. So the ninja overlay is nuked, which is in your stack trace and probably referenced by _skbuild. 

🤯

**Comment by joakim-hove** [comment]
> Maybe --no-build-isolation fixes this?

yes - that "fixes" it; after first wiping `_skbuild/` I can repeatedly call `pip install -e . --no-build-isolation` without wiping `_skbuild/` for every build.

**Comment by jondequinor** [comment]
With the quotes around fixes, do you mean it is not a proper fix? In my view, scikit-build does not support build isolation.

**Comment by joakim-hove** [comment]
> With the quotes around fixes, do you mean it is not a proper fix? In my view, scikit-build does not support build isolation.

Well - I am not familiar with `pip` beyond basic usage and I am even less familiar with `scikit-build`, but my view as an "end-developer" in this case is that is a bit weird with a build system which requires special flags like `--no-build-isolation` in order to be idempotent.

**Comment by jondequinor** [comment]
I agree, especially since https://github.com/scikit-build/scikit-build/issues/352 seems to make the workaround redundant. Maybe we're using skbuild wrong?

**Comment by jondequinor** [comment]
tl;dr: keep this issue open

**Comment by eivindjahren** [comment]
The current situation is that there are still situations in which you get failed builds and have to remove `_skbuild` although this will normally not happen. The issue will be resolved by https://github.com/equinor/ert/issues/6270

**Comment by oyvindeide** [comment]
Because the issue will be solved by removing the C code, I will close this issue as we will not action on it directly.
